[Enhancement] Introduce GC control by cluster snapshot info

srlch opened the pull request:
## Why I'm doing:
For the GC of cluster snapshot, we retain all version large than latest version of the current cluster snapshot createTime
But the problem is that too many version can not be vaccumed even the files does not contained in the snapshot version.

# What I'm doing:
Introduce cluster snapshot info based GC control for cluster snapshot as following:

1. Can not vacuum all retain versions from the cluster snapshot info.
2. Can not drop the partition/table/db in catalogbin if it is in cluster snapshot info.
3. Can not remove the data when doing StarMgrSyncer::deleteUnusedShardAndShardGroup if the corressponding index id is in cluster snapshot info.
4. Can not remove the data when doing StarMgrSyncer::syncTableMetaInternal if the corressponding index id is in cluster snapshot info.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshot.java:
```diff
@@ -42,6 +47,10 @@ public enum ClusterSnapshotType {
     @SerializedName(value = "starMgrJournal")
     private long starMgrJournalId;
 
+    // (DB id, table id, partition id) -> List(visible_version)
+    @SerializedName(value = "estimatedSnapshotVersions")
+    private Map<Pair<Long, Pair<Long, Long>>, List<Long>> estimatedSnapshotVersions;
+
```
better not to use pair, the code is not easy to maintain

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotCheckpointScheduler.java:
```diff
@@ -157,6 +188,37 @@ protected void runCheckpointScheduler() {
         }
     }
 
+    protected void captureSnapshotVersionInfo(Map<Pair<Long, Pair<Long, Long>>, List<Long>> snapshotVersionInfo) {
+        List<Long> dbIds = GlobalStateMgr.getCurrentState().getLocalMetastore().getDbIds();
+        for (Long dbId : dbIds) {
+            Database db = GlobalStateMgr.getCurrentState().getLocalMetastore().getDb(dbId);
+            if (db == null) {
+                continue;
+            }
+
+            List<Table> tables = new ArrayList<>();
+            for (Table table : GlobalStateMgr.getCurrentState().getLocalMetastore().getTables(db.getId())) {
+                if (table.isCloudNativeTableOrMaterializedView()) {
+                    tables.add(table);
+                }
+            }
+
+            for (Table table : tables) {
+                Locker locker = new Locker();
+                OlapTable olapTable = (OlapTable) table;
+                locker.lockTablesWithIntensiveDbLock(db.getId(), Lists.newArrayList(table.getId()), LockType.READ);
+                try {
+                    for (PhysicalPartition partition : olapTable.getPhysicalPartitions()) {
+                        Pair<Long, Pair<Long, Long>> key = Pair.create(dbId, Pair.create(table.getId(), partition.getId()));
+                        snapshotVersionInfo.computeIfAbsent(key, k -> Lists.newArrayList()).add(partition.getVisibleVersion());
+                    }
+                } finally {
+                    locker.unLockTablesWithIntensiveDbLock(db.getId(), Lists.newArrayList(table.getId()), LockType.READ);
+                }
+            }
+        }
+    }
+
```
should capture versions in checkpointing, not the serving global state

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshot.java:
```diff
@@ -42,6 +47,10 @@ public enum ClusterSnapshotType {
     @SerializedName(value = "starMgrJournal")
     private long starMgrJournalId;
 
+    // (DB id, table id, partition id) -> List(visible_version)
+    @SerializedName(value = "estimatedSnapshotVersions")
+    private Map<Pair<Long, Pair<Long, Long>>, List<Long>> estimatedSnapshotVersions;
+
```
it is better to not introduce this versions in automated snapshot ? only introduce it in manual snapshot ?

srlch replied:
updated

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -5247,4 +5248,40 @@ public Map<String, Long> estimateCount() {
         return ImmutableMap.of("Partition", totalCount);
     }
+
+    public void getNativeTableVersions(ClusterSnapshotInfo clusterSnapshotInfo) {
```
Not a good code style for java, this is C++ style.

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -5247,4 +5248,40 @@ public Map<String, Long> estimateCount() {
+
+            for (Table table : tables) {
+                Locker locker = new Locker();
+                locker.lockTablesWithIntensiveDbLock(db.getId(), Lists.newArrayList(table.getId()), LockType.READ);
```
No need to lock for checkpoint

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -5247,4 +5248,35 @@ public Map<String, Long> estimateCount() {
         return ImmutableMap.of("Partition", totalCount);
     }
+
+    public ClusterSnapshotInfo getNativeTableVersions() {
```
```suggestion
    public ClusterSnapshotInfo getClusterSnapshotInfo() {
```

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshot.java:
```diff
@@ -41,6 +41,8 @@ public enum ClusterSnapshotType {
     private long feJournalId;
     @SerializedName(value = "starMgrJournal")
     private long starMgrJournalId;
+    @SerializedName(value = "clusterSnapshotInfo")
+    private ClusterSnapshotInfo clusterSnapshotInfo;
 
```
Is it better to add a sub class ManualClusterSnapshot to include clusterSnapshotInfo?

xiangguangyxg commented on be/src/storage/lake/vacuum.cpp:
```diff
@@ -281,17 +281,30 @@ void run_clear_task_async(std::function<void()> task) {
 static Status collect_garbage_files(const TabletMetadataPB& metadata, const std::string& base_dir,
-                                    AsyncFileDeleter* deleter, int64_t* garbage_data_size) {
+                                    AsyncFileDeleter* deleter, int64_t* garbage_data_size,
+                                    const std::unordered_set<std::string>* retain_files_set) {
```
```suggestion
                                    const std::unordered_set<std::string>& retain_files_set) {
```

srlch replied:
updated

xiangguangyxg commented on be/src/storage/lake/vacuum.cpp:
```diff
@@ -429,11 +448,43 @@ static void erase_tablet_metadata_from_metacache(TabletManager* tablet_mgr, cons
 }
 
+static Status get_retain_files(const std::vector<int64_t>& retain_versions, int64_t tablet_id,
+                               TabletManager* tablet_mgr, std::unordered_set<std::string>& retain_files_set) {
```
```suggestion
                               TabletManager* tablet_mgr, std::unordered_set<std::string>* retain_files_set) {
```

xiangguangyxg replied:
input args use const reference, output args use pointer

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotCheckpointScheduler.java:
```diff
@@ -107,6 +110,14 @@ protected void runCheckpointScheduler() {
                 errMsg = "checkpoint journal id for FE is smaller than image version";
                 break;
             }
+            long finishTime = System.currentTimeMillis();
+            // Check the time taken to obtain consistency information. Reject the snapshot if
+            // it takes too long to finish to prevent the verion has been vacuumed.
+            if (job.needClusterSnapshotInfo() &&
+                    (finishTime - startTime) / 1000 > Config.lake_autovacuum_grace_period_minutes * 60) {
+                errMsg = "Take too long to obtain consistency information";
+                break;
+            }
```
could protect versions from being vacuumed during snapshoting ?

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/leader/CheckpointController.java:
```diff
@@ -96,14 +97,19 @@ public class CheckpointController extends FrontendDaemon {
         this.belongToGlobalStateMgr = Strings.isNullOrEmpty(subDir);
         nodesToPushImage = new HashSet<>();
+        this.needClusterSnapshotInfo = false;
+        this.clusterSnapshotInfo = new ClusterSnapshotInfo();
```
```suggestion
        this.clusterSnapshotInfo = null;
```

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/journal/CheckpointWorker.java:
```diff
@@ -40,20 +41,23 @@ public abstract class CheckpointWorker extends FrontendDaemon {
     private final AtomicReference<NextPoint> nextPoint = new AtomicReference<>();
     protected GlobalStateMgr servingGlobalState;
+    // every time we begin creating image, it will be reset
+    protected ClusterSnapshotInfo clusterSnapshotInfo = new ClusterSnapshotInfo();
```
```suggestion
    protected ClusterSnapshotInfo clusterSnapshotInfo;
```

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/journal/CheckpointWorker.java:
```diff
@@ -79,29 +84,30 @@ protected void runAfterCatalogReady() {
             return;
         }
 
+        this.clusterSnapshotInfo.reset();
```
```suggestion
        this.clusterSnapshotInfo = null;
```

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/leader/CheckpointController.java:
```diff
@@ -297,12 +306,15 @@ private String selectWorker() {
 
     private boolean doCheckpoint(Frontend frontend) {
+        // reset the cluster snapshot info before sending the checkpoint request
+        this.clusterSnapshotInfo.reset();
```
```suggestion
        this.clusterSnapshotInfo = null;
```

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotInfo.java:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    public void putPhysicalPartInfo(long dbId, long tableId, long partId, long physicalPartId, long version,
+                                    List<MaterializedIndex> indexes) {
```
```suggestion
                                    List<Long> indexIds) {
```

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/vacuum/AutovacuumDaemon.java:
```diff
@@ -102,6 +102,10 @@ public boolean shouldVacuum(PhysicalPartition partition) {
+        if (GlobalStateMgr.getCurrentState().getClusterSnapshotMgr().runningManualJob()) {
+            return false;
+        }
+
```
Fixed

srlch commented on fe/fe-core/src/main/java/com/starrocks/journal/CheckpointWorker.java:
```diff
@@ -40,20 +41,23 @@ public abstract class CheckpointWorker extends FrontendDaemon {
     private final AtomicReference<NextPoint> nextPoint = new AtomicReference<>();
     protected GlobalStateMgr servingGlobalState;
+    // every time we begin creating image, it will be reset
+    protected ClusterSnapshotInfo clusterSnapshotInfo = null;
```
Fixed

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:
```diff
@@ -301,6 +301,30 @@ public void removeAutomatedFinalizeJobs(int removeCount) {
+        return false;
+    }
+
+    public boolean checkIndexExistsInClusterSnapshotInfo(long dbId, long tableId, long partId, long indexId) {
```
```suggestion
    public boolean isMaterializedIndexInClusterSnapshotInfo(long dbId, long tableId, long partId, long indexId) {
```

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotInfo.java:
```diff
@@ -0,0 +1,189 @@
+        return physicalPartInfo.version;
+    }
+
+    public boolean checkDbExists(long dbId) {
```
```suggestion
    public boolean isDbExisted(long dbId) {
```

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotInfo.java:
```diff
@@ -0,0 +1,189 @@
+        return getDbInfo(dbId) != null;
+    }
+
+    public boolean checkTableExists(long dbId, long tableId) {
```
```suggestion
    public boolean isTableExisted(long dbId, long tableId) {
```

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotInfo.java:
```diff
@@ -0,0 +1,189 @@
+        return getTableInfo(dbId, tableId) != null;
+    }
+
+    public boolean checkPartitionExists(long dbId, long tableId, long partId) {
```
```suggestion
    public boolean isPartitionExisted(long dbId, long tableId, long partId) {
```

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotInfo.java:
```diff
@@ -0,0 +1,189 @@
+        return getPartitionInfo(dbId, tableId, partId) != null;
+    }
+
+    public boolean checkIndexExists(long dbId, long tableId, long partId, long indexId) {
```
```suggestion
    public boolean isMaterializedIndexExisted(long dbId, long tableId, long partId, long indexId) {
```

xiangguangyxg commented on be/src/storage/lake/vacuum.cpp:
```diff
@@ -451,12 +519,17 @@ static Status vacuum_tablet_metadata(TabletManager* tablet_mgr, std::string_view
     }
     int64_t final_vacuum_version = std::numeric_limits<int64_t>::max();
     for (auto& tablet_info : tablet_infos) {
+        std::unordered_set<std::string> retain_files_set;
```
do not use ?

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshot.java:
```diff
@@ -94,6 +104,18 @@ public long getId() {
+        return;
+    }
+
+    public long getSnapshotVersion(long dbId, long tableId, long partId, long physicalPartId) {
+        return 0;
+    }
```
not needed ?

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotCheckpointScheduler.java:
```diff
@@ -37,20 +37,17 @@ public class ClusterSnapshotCheckpointScheduler extends FrontendDaemon {
-
     @Override
     protected void runAfterCatalogReady() {
         if (!GlobalStateMgr.getCurrentState().getClusterSnapshotMgr().isAutomatedSnapshotOn()) {
```
shoud remove ?

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/vacuum/AutovacuumDaemon.java:
```diff
@@ -197,6 +197,16 @@ private void vacuumPartitionImpl(Database db, OlapTable table, PhysicalPartition
+                        GlobalStateMgr.getCurrentState().getClusterSnapshotMgr()
+                                      .getRetainVersionsForVacuum(
+                                            db.getId(), table.getId(), partition.getParentId(), partition.getId());
+        if (snapshotRetainVersions == null) {
+            LOG.info("Skip vacuum for {}.{}.{}.{}, because the retain versions info is not ready",
+                     db.getFullName(), table.getName(), partition.getParentId(), partition.getId());
+            return;
+        }
```
do not need to skip ?

srlch replied:
updated

alvin-celerdata commented on be/src/storage/lake/tablet_retain_info.h:
```diff
@@ -0,0 +1,82 @@
+    TabletRetainInfo() = default;
+    ~TabletRetainInfo() = default;
+
+    Status build_info(const std::vector<int64_t>& retain_versions, int64_t tablet_id, TabletManager* tablet_mgr) {
```
All implementations should be in C++ files. Only performance-critical code can be allowed in header files.

alvin-celerdata commented on be/src/storage/lake/tablet_retain_info.h:
```diff
@@ -0,0 +1,82 @@
+#include <unordered_set>
+#include <vector>
+
+#include "storage/lake/tablet_metadata.h"
```
If code is moved to c++ files, this is not necessary in header file.

alvin-celerdata commented on gensrc/proto/lake_service.proto:
```diff
@@ -297,6 +297,8 @@ message VacuumRequest {
     repeated TabletInfoPB tablet_infos = 7;
     // Whether enable file bundling
     optional bool enable_file_bundling = 8;
+    // retain versions info
```
What versions are they? There are several versions in SR. You should explain clearly what kind of version it is

srlch replied:
added more comment

srlch commented on be/src/storage/lake/tablet_retain_info.h:
```diff
@@ -0,0 +1,48 @@
+
+class TabletManager;
+
+class TabletRetainInfo {
```
https://github.com/StarRocks/starrocks/pull/59686 separate another PR to introduce this class and its UT with some comment in the header file

srlch commented on be/src/storage/lake/tablet_retain_info.h:
```diff
@@ -0,0 +1,48 @@
+
+    Status build_info(const std::vector<int64_t>& retain_versions, int64_t tablet_id, TabletManager* tablet_mgr);
+
+    bool filter_by_file_name(const std::string& file_name) const;
+
+    bool filter_by_version(int64_t version) const;
+
+    bool filter_by_rowset_id(uint32_t rowset_id) const;
```
fixed, use another name in the new PR

alvin-phoenix-ai reviewed (commented):
better to split it into small changes.

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/StarMgrMetaSyncer.java:
```diff
@@ -189,24 +189,27 @@ private void deleteUnusedShardAndShardGroup() {
+            long dbId = Long.parseLong(shardInfo.getLabels().get("dbId"));
+            long tableId = Long.parseLong(shardInfo.getLabels().get("tableId"));
+            long partitionId =  Long.parseLong(shardInfo.getLabels().get("partitionId"));
+            long indexId = Long.parseLong(shardInfo.getLabels().get("indexId"));
```
updated

srlch commented on be/src/storage/lake/vacuum.cpp:
```diff
@@ -548,6 +563,9 @@ static Status vacuum_tablet_metadata(TabletManager* tablet_mgr, std::string_view
         AsyncFileDeleter metafile_deleter(INT64_MAX, metafile_delete_cb);
         auto meta_dir = join_path(root_dir, kMetadataDirectoryName);
         for (auto v = vacuum_version_range->min_version; v < vacuum_version_range->max_version; v++) {
+            if (std::find(retain_versions.begin(), retain_versions.end(), v) != retain_versions.end()) {
```
Fixed

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/StarMgrMetaSyncer.java:
```diff
@@ -189,24 +189,36 @@ private void deleteUnusedShardAndShardGroup() {
+            if (GlobalStateMgr.getCurrentState().getClusterSnapshotMgr().isMaterializedIndexInClusterSnapshotInfo(
+                    dbId, tableId, partitionId, indexId)) {
+                LOG.debug("shard group {} can not be delete shard for now, because it exists in cluster snapshot info",
+                          shardGroupId);
```
Fixed

srlch committed:
[Enhancement] Introduce GC control by cluster snapshot info

For the GC of cluster snapshot, we retain all version large than latest version of the current cluster snapshot createTime
But the problem is that too many version can not be vaccumed even the files does not contained in the snapshot version.

Introduce cluster snapshot info based GC control for cluster snapshot as following:

1. Can not vacuum all retain versions from the cluster snapshot info.
2. Can not drop the partition/table/db in catalogbin if it is in cluster snapshot info.
3. Can not remove the data when doing StarMgrSyncer::deleteUnusedShardAndShardGroup if the corressponding index id is in cluster snapshot info.
4. Can not remove the data when doing StarMgrSyncer::syncTableMetaInternal if the corressponding index id is in cluster snapshot info.

Signed-off-by: srlch <linzichao@starrocks.com>

luohaha commented:
Not need to record segment files and del files when call `TabletRetainInfo::init`.

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>
